### PR TITLE
Fix can't change locked parameter names in full-app embedding

### LINKED ISSUE #25739: Locked parameters stick to the embedding on publish
**Describe the bug**
If you publish a question with a locked parameter and then change the parameter name in the query, Metabase will keep asking for the old parameter.

Workaround: Change the variable back to whatever JWT errors, save question, open embedding dialog and change the parameter to Editable, publish, close embedding dialog, change the variable, save question, open embedding dialog and set parameters as intended.

**To Reproduce**
1) enable embedding
2) create a SQL question with a parameter (e.g. select * from people where state = {state}), save it and publish it
3) then edit the SQL question and change the name for the parameter (e.g. select * from people where state = {some_other_state_parameter})
4) go to publish again the question, see that it's still expecting the first parameter

**Expected behavior**
The parameters need to be the ones that the question exposes

**Screenshots**
This question expects a "user", while "id" is the parameter that we can pass
image
image

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Metabase version: 1.44.4
 
**Severity**
P2

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by WiNloSt**
Fixes #25739

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/27179)

### CODE REVIEW

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx [hunk 1]
```diff
@@ -235,7 +235,20 @@ class EmbedModalContent extends Component {
...
+  const { resource, resourceParameters } = props;
+
+  return filterValidResourceParameters(
+    resource.embedding_params || {},
+    resourceParameters,
+  );
+}
+
+function filterValidResourceParameters(embeddingParams, resourceParameters) {
+  const validParameters = resourceParameters.map(parameter => parameter.slug);
```
Comment (WiNloSt) [inline frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:248]: the value that's set in `embedding_params` is from `parameter.slug`

https://github.com/metabase/metabase/blob/8c7d0bcdd6526201d5ab2349a42f8305fbfd52a8/frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx#L81

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx [hunk 2]
```diff
@@ -235,7 +235,20 @@ class EmbedModalContent extends Component {
...
+
+  return _.pick(embeddingParams, validParameters);
```
Comment (WiNloSt) [inline frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: The problem was that we don't remove the old parameter key when we rename the parameter, so let's say we rename the parameter name from `state` to `state_new` and published the `state` parameter before.

on `master` we'll save both `state` and `state_new` to `embedding_params` that's why the embedding panel still says it requires the old parameter's value (`state`) in this case.

File: frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx [hunk 3]
```diff
@@ -71,8 +69,11 @@ const AdvancedSettingsPane = ({
                 className="mr2"
                 style={{ color: color("text-light") }}
               />
-              <h3>{parameter.name}</h3>
+              <h3 id={`parameter-${parameter.slug}`}>{parameter.name}</h3>
               <Select
+                buttonProps={{
```
Comment (WiNloSt) [inline frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx:74]: Make this component more accessible and testable.

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 2 above)
Comment (kulyk) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: @WiNloSt should we do something to actually remove old parameters after a rename?

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        storeInitialState,
+        reducers: {
+          settings: (
+            state: Record<string, any> = storeInitialState.settings,
+          ) => {
+            return state;
+          },
+        },
+      },
+    );
```
Comment (kulyk) [inline frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:37]: WDYT about extracting a shared `setup` function or something that'd render the component and configure redux? I see identical steps duplicated across all the tests, so it's a bit hard to parse what exactly is different in each test

Comment (kulyk) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:37]: Also, this test doesn't have any assertions

**Review by kulyk — COMMENTED** [review]
Overall looks good; just wondering if we can avoid having outdated parameters in `embedding_params`. Also got a couple of comments on unit tests and big thanks for adding new ones!

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 2 above)
Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: Do you mean should we do something to remove old params that have been saved? If this PR is merged and they revisit the published question/dashboard and try to publish the question/dashboard again then the old param should be removed.

The only thing we could do to remove old parameters without involving user interactions is to migrate data on the BE. Not sure how far we want to go that route though.

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx (hunk 4 above)
Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:37]: Ahh yep, I forgot to extract those setup step thanks for calling that out.

Re: don't have any assertion. I'm not sure where I've seen this before but at the end of the day I just want to make sure we are able to render the component. But I think that we could still definitely assert certain elements that it's been render and that should be more explicit. 👍

Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx]: Fixed.
1. Extract repeating setup to `renderWithConfiguredProviders`, I don't like `setup` since it's too broad and doesn't sound like it would try to render anything imo.
2. Add some assertions to the first test.

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 2 above)
Comment (ranquild) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: Should we maybe remove invalid parameters also when publishing, e.g. here https://github.com/metabase/metabase/blob/9bfa9296c4e871b6ab1c57d7d3a9706f628b9730/frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx#L52?

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx [hunk 5]
```diff
@@ -235,7 +235,20 @@ class EmbedModalContent extends Component {
...
+function getDefaultEmbeddingParams(props) {
+  const { resource, resourceParameters } = props;
+
+  return filterValidResourceParameters(
+    resource.embedding_params || {},
+    resourceParameters,
+  );
+}
+
+function filterValidResourceParameters(embeddingParams, resourceParameters) {
```
Comment (ranquild) [inline frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:247]: nit: not sure if adding a separate function increases readability here

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 2 above)
Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: When publishing, `embeddingParams` should already be valid. This PR addressed the issue by ensuring the state `embeddingParams` is always valid since it's been validated when we populate the state https://github.com/metabase/metabase/blob/9bfa9296c4e871b6ab1c57d7d3a9706f628b9730/frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx#L44

After that there shouldn't be a case that `embeddingParams` could contain old parameters. That's why there's no need to specifically clean/validate `embeddingParams` anywhere else.

Hope that makes sense.

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 5 above)
Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:247]: This addresses the issue that the setup step keeps repeating the same options for `renderWithProvider` https://github.com/metabase/metabase/pull/27179#discussion_r1047530375

File: frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx (hunk 2 above)
Comment (kulyk) [reply frontend/src/metabase/public/components/widgets/EmbedModalContent.jsx:250]: Thanks for explaining what's going on here @WiNloSt, appreciate it! Looks good to me so far!

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx (hunk 3 above)
Comment (WiNloSt) [reply frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx:74]: Fixed
